openage/convert/opus/opusenc.pyx

```diff
@@ -0,0 +1,428 @@
+            memcpy(padbuf, inbuf, in_sz - in_pos)
+            memset(padbuf + in_sz - in_pos, 0, padbuf_sz - (in_sz - in_pos))
+            inbuf = padbuf
+
+        in_pos += nb_samples * wframe_len
+        ttl_samples += nb_samples
+
+
+        # convert.
+        enc_bytes = opus.opus_encode(oe, <opus.opus_int16 *> inbuf, frame_sz, op.packet, outbuf_sz)
+        # check how much was converted/ loop or EOF ?
+        if enc_bytes < 0:
+            free_buffers(&os, oe, padbuf, op.packet)
+            return "Encoding error in opus_encode(): {}".format(enc_bytes)
+
+        # append converted opuspacket.
+        op.bytes = enc_bytes
+        op.granulepos += nb_samples * (48000 / coding_rate)
+        if op.e_o_s:
+            pass  # TODO. set granulepos ? for resampling decoders.
+        op.packetno += 1
+
+        err = ogg.ogg_stream_packetin(&os, &op)
+        if err:
+            free_buffers(&os, oe, padbuf, op.packet)
+            return "ogg_stream_packetin() failed."
+
+        # Try to write page or force, if end of stream is reached.
+        while ogg.ogg_stream_pageout(&os, &og)\
+                or (op.e_o_s and ogg.ogg_stream_flush(&os, &og)):
+            outsize = write_ogg_page(outsize, outdata, &og)
+            if outsize == 0:
+                free_buffers(&os, oe, padbuf, op.packet)
+                raise MemoryError("Could not write opus data.")
+
+
+    free_buffers(&os, oe, padbuf, op.packet)
+
+    return outdata[:outsize]
+
+
+cdef int write_opus_header(ogg.ogg_stream_state *os, ogg.ogg_packet *op, inopt):
+    '''
+    Write the Opus header.
+    Returns non-zero on failure.
+    '''
+    try:
+        buf = bytearray(19)
+    except MemoryError:
+        return 1
+
+    buf[0:8] = b'OpusHead'
+    buf[8] = 1                # Opus encapsulating version.
+    buf[9] = inopt['channels']
+    buf[10:12] = inopt['pre_skip'].to_bytes(2, 'little')
+    buf[12:16] = inopt['input_rate'].to_bytes(4, 'little')
+    buf[16:18] = b'\x00\x00'  # Output gain
+    buf[18] = 0               # Channel mapping family
+
+
+    op.packet = buf
+    op.bytes = <long> 19
+    op.b_o_s = <long> 1
+    op.e_o_s = <long> 0
+    op.granulepos = <ogg.ogg_int64_t> 0
+    op.packetno = <ogg.ogg_int64_t> 0
+
+    return ogg.ogg_stream_packetin(os, op)
+
+
+cdef int write_opus_comment(ogg.ogg_stream_state *os, ogg.ogg_packet *op):
+    '''
+    Write the Opus comment header.
+    Returns non-zero on failure.
+    '''
+    try:
+        vendor = b'openage asset converter'
+        buf = b''.join((b'OpusTags', len(vendor).to_bytes(1, 'little'), vendor,
+                        b'\x00\x00\x00\x00'))  # number of comments following
+    except MemoryError:
+        return 1
+
+    op.packet = buf
+    op.bytes = <long> len(buf)
+    op.b_o_s = <long> 0
+    op.e_o_s = <long> 0
+    op.granulepos = <ogg.ogg_int64_t> 0
+    op.packetno = <ogg.ogg_int64_t> 1
+
+    return ogg.ogg_stream_packetin(os, op)
+
+
+def read_wav(wav):
```

Comment:
Making it a `cdef` and using a few more of the pyx language extensions. Basically every cdef'd variable is handled by native C++ code without invoking the python interpreter, which is way faster.